microbenchmarks/src/main/java/org/apache/bookkeeper/client/DigestTypeBenchmark.java

```diff
@@ -0,0 +1,172 @@
+ * under the License.
+ *
+ */
+
+package org.apache.bookkeeper.client;
+
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+import io.netty.buffer.ByteBufAllocator;
+import org.apache.bookkeeper.client.BookKeeper.DigestType;
+import org.apache.bookkeeper.util.DoubleByteBuf;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+/**
+ * Microbenchmarks for different digest type
+ * getting started:
+ * 1. http://tutorials.jenkov.com/java-performance/jmh.html
+ * 2. http://hg.openjdk.java.net/code-tools/jmh/file/tip/jmh-samples/src/main/java/org/openjdk/jmh/samples/
+ * 3. google
+ *
+ * To run:
+ * build project from command line.
+ * execute ./run.sh
+ */
+public class DigestTypeBenchmark {
+
+    public enum BufferType {
+        ARRAY_BACKED,
+        NOT_ARRAY_BACKED,
+        BYTE_BUF_DEFAULT_ALLOC
+    }
+
+    public enum Digest {
+        MAC,
+        CRC32,
+        CRC32_C,
+    }
+
+    static byte[] randomBytes(int sz) {
+        byte[] b = new byte[sz];
+        ThreadLocalRandom.current().nextBytes(b);
+        return b;
+    }
+
+    @State(Scope.Thread)
+    public static class MyState {
+
+        @Param
+        public BufferType bufferType;
+        @Param
+        public Digest digest;
+
+        private DigestManager crc32;
+        private DigestManager crc32c;
+        private DigestManager mac;
+
+        private ByteBuf arrayBackedBuffer;
+        private ByteBuf notArrayBackedBuffer;
+        private ByteBuf byteBufDefaultAlloc;
+
+        public ByteBuf digestBuf;
+
+        @Setup(Level.Trial)
+        public void doSetup() throws Exception {
+            final byte[] password = "password".getBytes("UTF-8");
+            crc32 = DigestManager.instantiate(ThreadLocalRandom.current().nextLong(0, Long.MAX_VALUE),
+                    password, DigestType.CRC32);
+
+            crc32c = DigestManager.instantiate(ThreadLocalRandom.current().nextLong(0, Long.MAX_VALUE),
+                    password, DigestType.CRC32C);
+
+            mac = DigestManager.instantiate(ThreadLocalRandom.current().nextLong(0, Long.MAX_VALUE),
+                    password, DigestType.MAC);
+
+            digestBuf = Unpooled.buffer(getDigestManager(digest).getMacCodeLength());
+
+            final int entrySizeKb = 64;
```

Comment:
Might be worth comparing also with different entry sizes.

Reply:
will do